### ISSUE TITLE
Allow IPPools to be dedicated to specific workloads by annotation

### ISSUE THREAD

**Issue by jtackaberry** [projectcalico/calico#6787]
I am creating an IPPool specific to a workload and using the `cni.projectcalico.org/ipv[46]pools` annotation to assign the respective v4 and v6 IPPools to this workload.  This works well.

The problem is that IPAM considers this pool to be generally available for all workloads.   There's no apparent mechanism to prevent allocation from this pool while also allowing the workload containing the annotation to use it.  Setting `disabled: true` prevents the annotated workload from using it, as does creating an IPReservation to mask the range used by the workload-specific IPPool.

## Expected Behavior

New to Calico, my original intuition was that only the default IPPool created during deployment would be used for workloads unless explicitly annotated.  But I can certainly understand use cases benefiting from multiple general-purpose IPPools, so the current default behavior isn't unreasonable

## Current Behavior

IPPools are fair game for all workloads.  There doesn't seem to be a way to dedicate an IPPool to a specific workload.

## Possible Solution

There should be some means of preventing an IPPool from being consumed for general use while also enabling it for specific workloads.

One approach would be to add a pod selector (in addition to the current node selector) that directly maps the IPPool to a given workload.

More intuitive to me, though, is to continue to perform the assignment via annotation and simply add a new field to IPPool that prevents it from being used except for explicitly annotated pods.  Something like `annotationOnly: true` or some such.

## Steps to Reproduce (for bugs)

1. Create a small IPPool
2. Add a `cni.projectcalico.org/ipv[46]pools` annotation to some workload referencing this pool
3. Deploy workload
4. Restart pods in other workloads and observe the small pool is consumed by those other workloads

## Context

My use case is to ensure a predictable range of IPv6 addresses for egress connections from a specific K8s StatefulSet. The pods in question initiate outbound SMTP where forward and reverse DNS should be setup as a best practice.  My aim is to allocate a /125, preconfigure the AAAA and PTR records, and deploy the MTA under this specific range.  I also included a v4 IPPool for consistency but this is less important because IPv4 is NATed in this cluster.

## Your Environment

* Calico v3.21.4
* Kubernetes 1.24.4
* Ubuntu 22.04

**Comment by caseydavenport** [comment]
IPAM will use the pool because its nodeSelector is defaulted to `all()`.

One option is to modify your node selector in that pool so that it doesn't select any nodes in your cluster - this will ensure that no nodes use it as one of their "default" pools to allocate from. 

Obviously not an ideal solution, but it does work for now.

**Comment by caseydavenport** [comment]
A "nicer" solution would be to add a field to the pool - e.g., `disableAutoAllocation: false` - or similar.

**Comment by jtackaberry** [comment]
@caseydavenport thanks very much for the suggestion.  In fact, now that I look at the docs for nodeSelector, I see that this is [actually documented](https://projectcalico.docs.tigera.io/reference/resources/ippool#node-selector) so there's undoubtedly an element of user error here, having missed that.  `disableAutoAllocation` would certainly be clearer, though.

My main concern now is that as my workload-specific IPPool is defining a subset of the larger pod CIDR range, I'm not sure if the overlap is treated as a mask or reservation that would prevent allocation within that range from the larger pool. Is it theoretically possible for IPAM to auto-assign an IP from the large default pool that conflicts with the smaller workload IPPool?

**Comment by caseydavenport** [comment]
Ah, it sounds like you may have bypassed our validation to create two pools with overlapping CIDR ranges. Generally Calico assumes that IP pools do not overlap (and in fact, has validation to prevent this if you use the projectcalico.org/v3 API - more on that here: https://github.com/projectcalico/calico/issues/6412).

I don't know precisely what will happen with overlapping pools - I would guess that the allocation code might see the larger pool, see that it selects the node, and then happily assign from the reserved pool because it's not expecting an overlap to exist.

I think we'd need a bit more of an enhancement to support what you're trying to do here. Right now, the only way to do this would be to create distinct non-overlapping pools. I understand that for some topologies this can be rather tedious.

**Comment by jtackaberry** [comment]
> Ah, it sounds like you may have bypassed our validation to create two pools with overlapping CIDR ranges.

Indeed, I *am* directly creating a `crd.projectcalico.org/v1` IPPool resource rather than `projectcalico.org/v3` via calicoctl, because I'm creating it as part of a larger Helm chart.  (Unless there's a more correct way to create Calico resources within charts?)

It is a bit problematic to carve out a separate, non-overlapping CIDR range for this given my network, but I might be able to apply some elbow grease here.

I'll leave it to you to decide if this is a use case you want to support.  My main obstacle has been addressed via the `!all()` nodeSelector. 

Thanks!

**Comment by jtackaberry** [comment]
Use case still applicable

**Comment by caseydavenport** [comment]
IP pools do now contain an AssignmentMode field, which controls whether or not the pool is valid for automatic IP assignment or reserved for manual IP selection: https://github.com/projectcalico/calico/blob/5559ef6b9dd26aa65e501df49e0de680f3414536/api/pkg/apis/projectcalico/v3/ippool.go#L128-L134

Thanks to @MichalFupso for that one

**Comment by jtackaberry** [comment]
Wonderful!  I guess this is only available starting in 3.30.  I'm using Microk8s and they tend to lag behind a couple releases, but once they move to 3.30.x I will be happy to kick those tires.

Thanks @caseydavenport  and @MichalFupso!
